### DISCUSSION TITLE
我真的要成為軟體工程師???

### DISCUSSION THREAD

**Discussion by HsuSungTing** [1101-GS4719-C/Python-Community-GS4719-C#1162] [General]
大家好，我是電機系的徐松廷，相信大家來上這門課，多多少少都對software development有興趣，
也可能考慮過要在未來成為軟體工程師。筆者我也是中之一，目前正在迷惘，當初選電機主要是覺得萬一自己不適合coding，還是可以去做硬體(電機系最主要的好處)。
但在人生的十字路口上，還是得做出選擇的，所以我整理了成為軟體工程師所需的技能與人格特質，我們一起來看看吧。

技能：
1.programming language：需要精通至少一種編程語言，例如Java、Python、C++等等。對於不同的開發任務，適當的編程語言是非常重要的。
2.資料庫：軟體工程師需要了解資料庫的基本知識，包括如何設計和管理資料庫、如何優化查詢效能以及如何保護數據安全等。
3.敏捷開發：敏捷開發是一種開發模型，可以讓開發團隊更有效地合作，更快地發佈軟體產品。軟體工程師需要了解敏捷開發的基本原則和實踐方法。
4.軟體設計模式：軟體設計模式是指解決特定問題的通用方法。軟體工程師需要了解和應用不同的設計模式，以實現更好的代碼可讀性、重用性和可維護性。
5.測試能力：測試和調試是軟體開發過程中至關重要的步驟。軟體工程師需要了解如何進行單元測試、集成測試和系統測試，以及如何有效地調試代碼中的錯誤。

人格特質：
1.創造力：需要具備創造力，以創建新的軟體產品或解決現有的問題。創造力可以幫助軟體工程師想出更好的解決方案，並使他們更具創新性和靈活性。
2.專注力：開發軟體需要長時間的專注和集中注意力。軟體工程師需要具備專注力，以便在編寫代碼、調試代碼或測試時能夠保持高效率和高品質。
3.自學能力：需要持續學習新的技術和工具，以跟上行業的變化。成功的軟體工程師需要具備自學能力，以便獲得新知識和技能，從而不斷提高自己的競爭力。
4.團隊合作能力：軟體開發是一個團隊活動，需要軟體工程師與其他開發人員、設計師和項目經理合作。軟體工程師需要具備良好的溝通和協作能力，以便有效地與團隊成員互動並解決問題。
5.堅韌性：軟體開發過程中會遇到許多困難和挑戰，例如代碼錯誤、時間壓力、預算限制等等。成功的軟體工程師需要具備堅韌性，以便在面對挑戰時保持冷靜和積極。

看了整理好的資料後，我時常還是懷疑變化速度極快的資訊產業是否適合自己，我個人的話是對coding有興趣，也喜歡自己做一些tool來用，但都是興趣(在電機系算了一大堆電路之後，還是會想玩玩軟體當興趣，不然有時候好枯燥)，所以目前也沒把握真正成為軟體工程師是否適合自己，就先修課看看吧，且走且看。

以下歡迎大家分享自己的狀況或想法!

**Comment by xxrjun** [comment]
非常感謝同學的分享!

技能的部分我想補充**資料結構與演算法**以及**系統分析與設計**
- 運用好的結構與演算法可以使程式提升程式效率、降低成本，這兩個好比一個工程師的「內功」，也是資訊工程領域的必修課。(不過網頁前端工程師應該是不需要)
- 系統分析與設計涉及到如何通過設計合理的系統來滿足用戶需求，如何理解需求、規劃軟體開發流程、進行設計、開發與測試，以及如何在專案中與團隊成員協作等等。這些技能可以幫助開發人員開發出更好的軟體產品，並有效地管理和完成軟體開發項目。

然後我特別認同人格特質中第三點的**自學能力**! 科技與技術的變化非常快，需要持續學習新的技術和工具才有辦法在這個領域繼續待下去。我看一些工程師前輩分享自身經驗時也都很強調這一點，公司並不會永遠都使用舊有的技術，只要新技術出現並於評估後決定更換，工程師就需要盡快習得新技術，學得慢或無法跟上很快就會被替換掉。另外，能夠看懂官方文件或是新鮮的論文我覺得也是自學能力中很重要的一部分，過去我在學習時都是看國內技術分享的文章或是書籍，但是第二手資訊往往慢很多也常常不完整，為了能夠學習到最新的工具與技術，後來就慢慢培養出自己邊看官方文件邊學習開發的習慣與能力(持續培養學習中，順便偷臭 React 以前的官方文件偏:poop:)

**Comment by poiu6fi** [comment]
看完了上述需要的條件跟技能後，我突然我的我好像不是走這塊的料
